contracts/Redemption.vy

```diff
@@ -0,0 +1,495 @@
+owner: public(address)
+# @dev Returns the address of the pending owner.
+pending_owner: public(address)
+# @dev when the contract is killed, redemptions aren't possible
+killed: public(bool)
+# @dev recipient of the ETH used for redemptions
+payee: public(address)
+# @dev scaling factor parameters packed into a single slot
+packed_scaling_factor: uint256
+
+# @dev Emitted when contract is killed
+event Killed:
+    yfi_recovered: uint256
+
+event Sweep:
+    token: indexed(address)
+    amount: uint256
+
+# @dev Emitted when the ownership transfer from
+# `previous_owner` to `pending_owner` is initiated.
+event PendingOwnershipTransfer:
+    previous_owner: indexed(address)
+    pending_owner: indexed(address)
+
+
+# @dev Emitted when the ownership is transferred
+# from `previous_owner` to `new_owner`.
+event OwnershipTransferred:
+    previous_owner: indexed(address)
+    new_owner: indexed(address)
+
+event SetPayee:
+    payee: indexed(address)
+
+
+MASK: constant(uint256) = 2**64 - 1
+
+# powers of 10
+E3: constant(int256)               = 1_000
+E6: constant(int256)               = E3 * E3
+E9: constant(int256)               = E3 * E6
+E12: constant(int256)              = E3 * E9
+E15: constant(int256)              = E3 * E12
+E17: constant(int256)              = 100 * E15
+E18: constant(int256)              = E3 * E15
+E20: constant(int256)              = 100 * E18
+MIN_NAT_EXP: constant(int256)      = -41 * E18
+MAX_NAT_EXP: constant(int256)      = 130 * E18
+
+# x_n = 2^(7-n), a_n = exp(x_n)
+# in 20 decimals for n >= 2
+X0: constant(int256)  = 128 * E18 # 18 decimals
+A0: constant(int256)  = 38_877_084_059_945_950_922_200 * E15 * E18 # no decimals
+X1: constant(int256)  = X0 / 2 # 18 decimals
+A1: constant(int256)  = 6_235_149_080_811_616_882_910 * E6 # no decimals
+X2: constant(int256)  = X1 * 100 / 2
+A2: constant(int256)  = 7_896_296_018_268_069_516_100 * E12
+X3: constant(int256)  = X2 / 2
+A3: constant(int256)  = 888_611_052_050_787_263_676 * E6
+X4: constant(int256)  = X3 / 2
+A4: constant(int256)  = 298_095_798_704_172_827_474 * E3
+X5: constant(int256)  = X4 / 2
+A5: constant(int256)  = 5_459_815_003_314_423_907_810
+X6: constant(int256)  = X5 / 2
+A6: constant(int256)  = 738_905_609_893_065_022_723
+X7: constant(int256)  = X6 / 2
+A7: constant(int256)  = 271_828_182_845_904_523_536
+X8: constant(int256)  = X7 / 2
+A8: constant(int256)  = 164_872_127_070_012_814_685
+X9: constant(int256)  = X8 / 2
+A9: constant(int256)  = 128_402_541_668_774_148_407
+X10: constant(int256) = X9 / 2
+A10: constant(int256) = 11_331_4845_306_682_631_683
+X11: constant(int256) = X10 / 2
+A11: constant(int256) = 1_064_49_445_891_785_942_956
+
+@external
+def __init__(
+    yfi: address, d_yfi: address, ve_yfi: address, owner: address, 
+    price_feed: address, curve_pool: address, scaling_factor: uint256,
+):
+    YFI = ERC20(yfi)
+    DYFI = IDYFI(d_yfi)
+    VEYFI = ERC20(ve_yfi)
+    PRICE_FEED = AggregatorV3Interface(price_feed)
```

Comment:
do we have the address to user for this, just want to check integration is correct as targeted

Reply:
Actually reminds me to copy in the YFI/ETH oracle that was written for yDiscount. It uses both YFI/USD and YFI/ETH chainlink oracles and multiplies them to get YFI/ETH price. This is done because the official YFI/ETH oracle has a 24h heartbeat.

The oracles are:
- YFI/USD: `0xA027702dbb89fbd58938e4324ac03B58d812b0E1`
- ETH/USD: `0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419`